zebra-state/src/service.rs

```diff
@@ -618,98 +612,21 @@ impl StateService {
-            metrics::gauge!(
-                "state.full_verifier.committed.block.height",
-                tip_block_height.0 as f64,
+            let finalized_tip_height = self.read_service.db.finalized_tip_height().expect(
+                "Finalized state must have at least one block before committing non-finalized state",
             );
 
-            // This height gauge is updated for both fully verified and checkpoint blocks.
-            // These updates can't conflict, because the state makes sure that blocks
-            // are committed in order.
-            metrics::gauge!(
-                "zcash.chain.verified.block.height",
-                tip_block_height.0 as f64,
-            );
+            self.queued_non_finalized_blocks
+                .prune_by_height(finalized_tip_height);
         }
 
-        tracing::trace!("finished processing queued block");
         rsp_rx
     }
 
-    /// Update the [`LatestChainTip`], [`ChainTipChange`], and `non_finalized_state_sender`
-    /// channels with the latest non-finalized [`ChainTipBlock`] and
-    /// [`Chain`][1].
-    ///
-    /// Returns the latest non-finalized chain tip height, or `None` if the
-    /// non-finalized state is empty.
-    ///
-    /// [1]: non_finalized_state::Chain
-    //
-    // TODO: remove this clippy allow when we remove self.chain_tip_sender
-    #[allow(clippy::unwrap_in_result)]
-    #[instrument(level = "debug", skip(self))]
-    fn update_latest_chain_channels(&mut self) -> Option<block::Height> {
-        let best_chain = self.mem.best_chain();
-        let tip_block = best_chain
-            .and_then(|chain| chain.tip_block())
-            .cloned()
-            .map(ChainTipBlock::from);
-        let tip_block_height = tip_block.as_ref().map(|block| block.height);
-
-        // If the final receiver was just dropped, ignore the error.
-        let _ = self.non_finalized_state_sender.send(self.mem.clone());
-
-        self.chain_tip_sender
-            .lock()
-            .expect("unexpected panic in block commit task or state")
-            .set_best_non_finalized_tip(tip_block);
-
-        tip_block_height
-    }
-
-    /// Run contextual validation on the prepared block and add it to the
-    /// non-finalized state if it is contextually valid.
-    #[tracing::instrument(level = "debug", skip(self, prepared))]
-    fn validate_and_commit(&mut self, prepared: PreparedBlock) -> Result<(), CommitBlockError> {
-        self.check_contextual_validity(&prepared)?;
-        let parent_hash = prepared.block.header.previous_block_hash;
-
-        if self.disk.db.finalized_tip_hash() == parent_hash {
-            self.mem.commit_new_chain(prepared, &self.disk.db)?;
-        } else {
-            self.mem.commit_block(prepared, &self.disk.db)?;
-        }
-
-        Ok(())
-    }
-
     /// Returns `true` if `hash` is a valid previous block hash for new non-finalized blocks.
     fn can_fork_chain_at(&self, hash: &block::Hash) -> bool {
```

Comment:
Required Fix: `can_fork_chain_at()` is used to commit ready non-finalized blocks, but it does not currently work, because `self.mem` is always empty.

Instead:
- change `last_block_hash_sent` into a local set of the block hashes that have recently been sent, but haven't made it into the non-finalized or finalized state yet
- because it's a set, we don't need to reset `last_block_hash_sent` if there is a block commit error, but we do need to discard outdated hashes (below the finalized tip)
- use `StateService.read_service.non_finalized_state_receiver` to get a cached `NonFinalizedState`

Let me know if you'd like me to push a PR with these changes.

Reply:
Here's how we use `OrderedMap` in the address book:
https://github.com/ZcashFoundation/zebra/blob/97fb85dca9d00a717661ea0cbab7ce66b8d23977/zebra-network/src/address_book.rs#L130

https://github.com/ZcashFoundation/zebra/blob/97fb85dca9d00a717661ea0cbab7ce66b8d23977/zebra-network/src/address_book.rs#L331-L337

https://github.com/ZcashFoundation/zebra/blob/97fb85dca9d00a717661ea0cbab7ce66b8d23977/zebra-network/src/address_book.rs#L397-L403

Reply:
Sent hashes added in [`03b2a2f`](https://github.com/ZcashFoundation/zebra/pull/5257/commits/03b2a2f4f693be7f4094020bf9be475b86e4e9c2)

I used a VecDeque for pruning by height because it looked like OrderedMap iterates over the entire set for insert and remove ops. It doesn't have a `contain` or `get` method either, so it would be doing that twice. 

Since hashes are being added to this set at the same time as they're sent to be committed to the non-finalized state, and not removed until after they're below the finalized tip, I removed the check against the non-finalized chains in `can_fork_chain_at`.

Reply:
`sent_non_finalized_block_hashes.contains()` can return `true` for blocks that are:
- below the finalised tip
- at the finalised tip height, but not the same hash as the finalised tip block

This will cause an error when committing those blocks.

There are a few different ways we could fix this, here are some alternatives:
- make pruning reliable
- pass the finalised tip height to `contains()`, and return false for any height that is less than or equal to the finalised tip

There's still a race between the finalised tip and the channel here, but it will be much smaller after we make this change. If the race does cause an error, we should make sure it's not a panic.

Reply:
I updated SentHashes to make pruning reliable in [`70beeb0`](https://github.com/ZcashFoundation/zebra/commit/70beeb07459d35940e4806342bba8daf48816d9c), it still may return true for blocks that are at the finalized tip height but not the same hash as the finalized tip block.